modules/nlu/src/backend/application/training-repo.ts

```diff
@@ -0,0 +1,138 @@
+import * as sdk from 'botpress/sdk'
+import { Transaction } from 'knex'
+import ms from 'ms'
+import { TrainingId, TrainingState, TrainingSession, I } from './typings'
+
+const TABLE_NAME = 'nlu_training_queue'
+const TRANSACTION_TIMEOUT_MS = ms('5s')
+const debug = DEBUG('nlu').sub('database')
+
+const TRANSACTION_TIMEOUT_ERROR = "Transaction exceeded it's time limit"
+
+const timeout = (ms: number) => {
+  return new Promise((_, reject) => {
+    setTimeout(() => reject(new Error(TRANSACTION_TIMEOUT_ERROR)), ms)
+  })
+}
+
+export type ITransactionContext = I<TransactionContext>
+class TransactionContext {
+  constructor(protected _database: typeof sdk.database, public transaction: Transaction | null = null) {}
+
+  private get table() {
+    if (this.transaction) {
+      return this._database.table(TABLE_NAME).transacting(this.transaction)
+    }
+    return this._database.table(TABLE_NAME)
+  }
+
+  public set = async (trainId: TrainingId, trainState: TrainingState): Promise<void> => {
+    const { botId, language } = trainId
+    const { progress, status, owner } = trainState
+
+    const modifiedOn = this._database.date.now()
+    if (await this.has({ botId, language })) {
+      return this.table.where({ botId, language }).update({ progress, status, modifiedOn, owner })
+    }
+    return this.table.insert({ botId, language, progress, status, modifiedOn, owner })
+  }
+
+  public has = async (trainId: TrainingId): Promise<boolean> => {
+    const { botId, language } = trainId
+    const result = !!(await this.get({ botId, language }))
+    return result
+  }
+
+  public get = async (trainId: TrainingId): Promise<TrainingSession | undefined> => {
+    const { botId, language } = trainId
+
+    return this.table
+      .where({ botId, language })
+      .select('*')
+      .first()
+  }
+
+  public getAll = async (): Promise<TrainingSession[]> => {
```

Comment:
is this used ? seems like dead to me.

Reply:
Used by `getAllTrainings` in `trainingQueue` which is then used in `getAllTrainings` in`NLUApplication` which is then only used for testing.

`NLUApplication` exposes the training repo, we could simply do `app.trainRepository.getAll()` and remove the  two extra `getAllTrainings`!